tree/ntuple/v7/doc/specifications.md

```diff
@@ -74,36 +103,63 @@ The meta-data envelope defines additional basic types (see below).
 
 ### Feature Flags
 
-Feature flags are 64bit integers where every bit represents a certain feature that is used
+Feature flags are 64bit integers where every bit represents a certain forward-incompatible feature that is used
 in the binary format of the RNTuple at hand.
 The most significant bit is used to indicate that there are more than 63 features to specify.
 That means that readers need to continue reading feature flags as long as their signed integer value is negative.
 
+Readers should gracefully abort reading when they encounter unknown bits set.
+
+The following feature bits are defined:
+
+| Bit                                | Feature              |
+------------------------------------ |----------------------|
+| 137 (0x09 of the 3rd feature int)  | Reserved for testing |
+
 
 ## Frames
 
 RNTuple envelopes can store records and lists of basic types and other records or lists by means of **frames**.
-The frame has the following format
+Frames can be short (<1GB) or long (>=1GB).
+A short frame has the following format
 
 ```
  0                   1                   2                   3
  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|                             Size                            |T|
+|0|                           Size                            |T|
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|              Number of Items (for list frames)                |
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                         FRAME PAYLOAD                         |
+|                              ...                              |
+```
+
+A long frame has the following format
+```
+ 0                   1                   2                   3
+ 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
-|           Number of Items (for list frames)           |Reserv.|
+|1|                                                             |
++-+                           Size                            +-+
+|                                                             |T|
++-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|              Number of Items (for list frames)                |
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 |                         FRAME PAYLOAD                         |
 |                              ...                              |
 ```
 
-_Size_: The size in bytes of the frame and the payload
+_Size_: The size in bytes of the frame and the payload.
+To get the size, first read a 32bit signed integer.
+An even integer indicates a short frame.
+An odd integer indicates a long frame; in this case, read a 64bit signed integer.
```

Comment:
I don't know the answer by is it really worth saving 4 bytes here to introduce a speed bump? (defined as one needs to read byte and stop and read bytes differently depending on value at runtime).

Places where this can matter:
- reading over network
- Implementing parsing in SIMT model (future GPU/IPU reading?)

Reply:
I can see complexity for the readers, but not necessarily the speed bump. The frames only occur in meta-data envelopes (header, footer, page list), which are compressed. So one would anyway read the entire compressed envelope, decompress, verify the envelope checksum, and then parse the contents.

We use frames extensively, so it felt worthwhile to save the 4 bytes for the vast majority of frames that are small. Let me see if I can get some better data on this.

Reply:
Looking into it, I think you are right. It's not worth the effort. The uncompressed meta-data size can grow up to 10% (nanoAOD) with static large frames, compressed the difference is negligible. A static frame size removes an issue during writing, where for large frames the payload needed to be `memmove`d by 4 bytes to finalize the frame.